File: Python/Leetcode/Matrix/setMatrixZeroes.py
class Solution():
    def setZeroes(self, matrix: list[list[int]]) -> None:
        indexFound = []

        # Finding the indices of zeros in the matrix
        for i in range(len(matrix)):
            for j in range(len(matrix[i])):
                if matrix[i][j] == 0:
                    indexFound.append([i, j])

        # Setting the corresponding rows and columns containing zeros to zero
        for item in indexFound:
            row_index, col_index = item[0], item[1]

            # Set entire row to zero
            matrix[row_index] = [0] * len(matrix[row_index])

            # Set entire column to zero
            for row in matrix:
                row[col_index] = 0

        return matrix
        
          
if __name__ == '__main__':
    solution_instance = Solution()
    print(solution_instance.setZeroes([[1,1,1],
                                 [1,0,1],
                                 [1,1,1]]))
    
    print(solution_instance.setZeroes([[0,1,2,0],
                                 [3,4,5,2],
                                 [1,3,1,5]]))

File: C++/Leetcode/Matrix/setMatrixZeroes.cpp
#include <iostream>
#include <vector>

class Solution {
    public:
        std::vector<std::vector<int>> setZeroes(std::vector<std::vector<int>> matrix){
            if (matrix.empty()); return;

            std::vector<std::vector<int>> res;    

            for (int i=0; i < matrix[0].size(); ++i){
                for (int y=0; y < matrix.size(); ++y){
                    if (matrix[i][y] == 0){
                        res.push_back({i,y});
                    }
                }
            }

            for (const auto& sector : res){
                matrix[sector[0]] = std::vector<int>(matrix[sector[0]].size(), 0);

                for (int k = 0; k < matrix[0].size(); ++k) {
                    matrix[sector[0]][k] = 0;
                }
                // Set entire column to zeros
                for (int k = 0; k < matrix.size(); ++k) {
                    matrix[k][sector[1]] = 0;
                }
            }

            return matrix;
        }
    private:
        void printList(std::vector<std::vector<int>> matrix){
            for (const auto& row : matrix){
                for (const auto& element : row){
                    std::cout << element << std::endl;
                }
            }
        }
};

int main(){
    Solution solution_instance;
    std::vector<std::vector<int>> matrix = {
        {1, 1, 1},
        {4, 0, 6},
        {7, 8, 9}
    };
    solution_instance.setZeroes(matrix);
    return 0;
}